pkg/template/processor.go

```diff
@@ -0,0 +1,156 @@
+	codecFactory serializer.CodecFactory
+}
+
+func NewProcessor(cl client.Client, scheme *runtime.Scheme) *Processor {
+	return &Processor{cl: cl, scheme: scheme, codecFactory: serializer.NewCodecFactory(scheme)}
+}
+
+func (p Processor) ProcessAndApply(tmplContent []byte, values map[string]string) error {
+	objs, err := p.Process(tmplContent, values)
+	if err != nil {
+		return errs.Wrap(err, "failed while processing template")
+	}
+
+	return p.apply(objs)
+}
+
+func (p Processor) apply(objs []runtime.RawExtension) error {
+	// sorting before applying to maintain correct order
+	sort.Sort(ByKind(objs))
+
+	for _, rawObj := range objs {
+		obj := rawObj.Object
+		if obj == nil {
+			continue
+		}
+		gvk := obj.GetObjectKind().GroupVersionKind()
+		if err := createOrUpdateObj(p.cl, obj); err != nil {
+			return errs.Wrapf(err, "unable to create resource of kind: %s, version: %s", gvk.Kind, gvk.Version)
+		}
+		if gvk.Group == "project.openshift.io" && gvk.Version == "v1" && gvk.Kind == "ProjectRequest" {
+			var prq projectv1.ProjectRequest
+			err := p.scheme.Convert(obj, &prq, nil)
+			if err != nil {
+				return errs.Wrapf(err, "unable to convert object of kind: %s, version: %s to a ProjectRequests", gvk.Kind, gvk.Version)
+			}
+			// wait until the Project exists and is ready
+			_, err = p.waitUntilProjectExists(prq.GetName(), 10*time.Second)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (p Processor) waitUntilProjectExists(namespace string, t time.Duration) (projectv1.Project, error) {
+	var prj projectv1.Project
+	ticker := time.NewTicker(t / 5)
+	defer ticker.Stop()
+	timeout := time.After(t)
+	for {
+		select {
+		case <-ticker.C:
+			err := p.cl.Get(context.TODO(), types.NamespacedName{
+				Namespace: namespace,
+				Name:      "",
+			}, &prj)
+			// something wrong happened while checking the project
+			if err != nil && !apierrors.IsNotFound(err) {
+				return projectv1.Project{}, errs.Wrapf(err, "failed to check if project '%s' exists", namespace)
+			}
+			// project exists and is active
+			if prj.Status.Phase == corev1.NamespaceActive {
+				return prj, nil
+			}
+			// project does not exist, so let's wait
+		case <-timeout:
+			return projectv1.Project{}, errs.Errorf("timeout while checking if project '%s' exists", namespace)
+		}
+	}
+}
+
+func createOrUpdateObj(cl client.Client, obj runtime.Object) error {
+	if err := cl.Create(context.TODO(), obj); err != nil {
+		if !apierrors.IsAlreadyExists(err) {
+			return errs.Wrapf(err, "failed to create object %v", obj)
+		}
+
+		if err = cl.Update(context.TODO(), obj); err != nil {
+			return errs.Wrapf(err, "failed to update object %v", obj)
+		}
+		return nil
+	}
+	return nil
+}
+
+// Process processes the template with the given content
+func (p Processor) Process(tmplContent []byte, values map[string]string) ([]runtime.RawExtension, error) {
+	decode := p.codecFactory.UniversalDeserializer().Decode
+	obj, _, err := decode(tmplContent, nil, nil)
```

Comment:
can this just be `obj, _, err :=p.codecFactory.UniversalDeserializer().Decode(tmplContent, nil, nil)`?

Reply:
sure, let me change that

Reply:
done in e4cc320